README.md

# Kalendar 1.0.0+3

Highly customizable Flutter Calendar.

| ![Image](https://raw.githubusercontent.com/bhanu-manthry/kalendar/master/screenshots/calendar_with_border.gif) | ![Image](https://raw.githubusercontent.com/bhanu-manthry/kalendar/master/screenshots/flutter_02.png) |
| :------------: | :------------: |
| **Kalendar** with border | **Kalendar** with custom day tiles |

| ![Image](https://raw.githubusercontent.com/bhanu-manthry/kalendar/master/screenshots/mark_selected_dates.gif) | ![Image](https://raw.githubusercontent.com/bhanu-manthry/kalendar/master/screenshots/mark_selected_dates_custom_day_tiles.gif) |
| :------------: | :------------: |
| Select multiple dates and mark them | Mark dates with custom events |

## Features

* Easy to use API
* Select single date
* Select multiple dates
* Mark dates with different types of events
* Build your own header, week row and day tiles with builders

## Usage
```dart
import 'package:flutter/material.dart';
import 'package:kalendar/kalendar.dart';

class SimpleCalendar extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.symmetric(horizontal: 4),
      child: Kalendar(),
    );
  }
}
```

## Props
| props                 | types                       |
| :---------------------| :-------------------------: |
| markedDates           | `Map<String, List<String>>` | 
| markBuilder           | `MarkBuilder`               |
| borderRadius          | `double`                    |
| onTap                 | `OnTappedOnDayTile`         |
| selectedDates         | `Set<String>`               |
| dayTileMargin         | `double`                    |
| dayTileBorderColor    | `Color`                     |
| dayTileBuilder        | `DayTileBuilder`            | 
| weekBuilder           | `WeekBuilder`               |
| showBorder            | `bool`                      |
| headerBuilder         | `HeaderBuilder`             |


CHANGELOG.md
## [0.0.1]

* First release.

## [1.0.0]

* Big fixes
* Improvements in API
* Update README.md


test/kalendar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:kalendar/kalendar.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  //   expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  // });
}


lib/kalendar.dart
library kalendar;

import 'package:flutter/material.dart';
import 'package:date_utils/date_utils.dart';

typedef ValueChanged<T> = void Function(T);
typedef MarkBuilder = Widget Function(String event);
typedef OnTappedOnDayTile = Widget Function(DateTime dateTime, bool isSelected);
typedef HeaderBuilder = Widget Function(
    DateTime dateTime, ValueChanged<DateTime> changedDateTime);
typedef DayTileBuilder = Widget Function(DayProps dayProps);
typedef WeekBuilder = Widget Function(List<String> weeks);

class Kalendar extends StatefulWidget {
  final Map<String, List<String>> markedDates;
  final MarkBuilder markBuilder;
  final double borderRadius;
  final double dayTileMargin;
  final OnTappedOnDayTile onTap;
  final Set<String> selectedDates;
  final Color dayTileBorderColor;
  final DayTileBuilder dayTileBuilder;
  final WeekBuilder weekBuilder;
  final bool showBorder;
  final HeaderBuilder headerBuilder;

  Kalendar({
    this.markedDates,
    this.markBuilder,
    this.borderRadius,
    this.onTap,
    this.selectedDates,
    this.dayTileMargin,
    this.dayTileBorderColor,
    this.dayTileBuilder,
    this.weekBuilder,
    this.showBorder,
    this.headerBuilder,
  });

  @override
  _KalendarState createState() => _KalendarState();
}

class _KalendarState extends State<Kalendar> {
  DateTime _visibleMonth;

  int _currentPageIndex = 999;
  PageController _pageController;

  _onPageChanged(pageIndex) {
    if (pageIndex > _currentPageIndex) {
      _visibleMonth = Utils.nextMonth(_visibleMonth);
    } else if (pageIndex < _currentPageIndex) {
      _visibleMonth = Utils.previousMonth(_visibleMonth);
    }

    setState(() {});

    _currentPageIndex = pageIndex;
  }

  void initState() {
    super.initState();
    _visibleMonth = Utils.firstDayOfMonth(DateTime.now());
    _pageController = PageController(initialPage: _currentPageIndex);
  }

  void _changeMonth(date) {
    setState(() {
      _visibleMonth = date;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        widget.headerBuilder != null
            ? widget.headerBuilder(_visibleMonth, _changeMonth)
            : _DefaultHeader(_visibleMonth, onChange: (date) {
                _changeMonth(date);
              }),
        widget.weekBuilder != null
            ? widget.weekBuilder(Utils.weekdays)
            : Container(
                padding: EdgeInsets.only(bottom: 12),
                child: Row(
                  children: List.generate(Utils.weekdays.length, (index) {
                    return Expanded(
                      child: Text(
                        '${Utils.weekdays[index]}',
                        textAlign: TextAlign.center,
                        style: TextStyle(color: Colors.red),
                      ),
                    );
                  }),
                ),
              ),
        Expanded(
          child: PageView.builder(
            controller: _pageController,
            onPageChanged: _onPageChanged,
            itemBuilder: (context, pageIndex) {
              List<DateTime> days = Utils.daysInMonth(_visibleMonth);
              return GridView.builder(
                  gridDelegate: new SliverGridDelegateWithFixedCrossAxisCount(
                      crossAxisCount: 7),
                  itemCount: days.length,
                  itemBuilder: (context, index) {
                    return Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: <Widget>[
                        DayTileContainer(
                          dateTime: days[index],
                          isDayOfCurrentMonth:
                              days[index].month == _visibleMonth.month,
                          events: widget.markedDates != null
                              ? widget.markedDates[formatDate(days[index])]
                              : null,
                          markBuilder: widget.markBuilder,
                          borderRadius: widget.borderRadius ?? 4,
                          dayTileMargin: widget.dayTileMargin,
                          onTap: widget.onTap,
                          isSelected: widget.selectedDates != null
                              ? widget.selectedDates
                                      .contains(formatDate(days[index])) ??
                                  false
                              : false,
                          dayTileBorderColor: widget.dayTileBorderColor,
                          dayTileBuilder: widget.dayTileBuilder,
                          showBorder: widget.showBorder,
                        ),
                      ],
                    );
                  });
            },
          ),
        ),
      ],
    );
  }
}

class _DefaultHeader extends StatelessWidget {
  final DateTime visibleMonth;
  final Function onChange;

  _DefaultHeader(this.visibleMonth, {this.onChange});

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.symmetric(vertical: 16, horizontal: 16),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.center,
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          IconButton(
            icon: Icon(
              Icons.chevron_left,
              color: Theme.of(context).primaryColor,
            ),
            onPressed: () {
              onChange(Utils.previousMonth(visibleMonth));
            },
          ),
          Text(
            '${Utils.formatMonth(visibleMonth)}',
            style: TextStyle(
              color: Theme.of(context).primaryColor,
              fontSize: 20,
              fontWeight: FontWeight.bold,
            ),
          ),
          IconButton(
            icon: Icon(
              Icons.chevron_right,
              color: Theme.of(context).primaryColor,
            ),
            onPressed: () {
              onChange(Utils.nextMonth(visibleMonth));
            },
          ),
        ],
      ),
    );
  }
}

class DayTileContainer extends StatelessWidget {
  final DateTime dateTime;
  final bool isDayOfCurrentMonth;
  final List<String> events;
  final Function markBuilder;
  final double borderRadius;
  final double dayTileMargin;
  final Function onTap;
  final bool isSelected;
  final Color dayTileBorderColor;
  final Function dayTileBuilder;
  final bool showBorder;

  DayTileContainer({
    this.dateTime,
    this.isDayOfCurrentMonth,
    this.events,
    this.markBuilder,
    this.borderRadius,
    this.dayTileMargin,
    this.onTap,
    this.isSelected,
    this.dayTileBorderColor,
    this.dayTileBuilder,
    this.showBorder,
  });

  @override
  Widget build(BuildContext context) {
    DayProps dayProps = DayProps(
      dateTime: dateTime,
      onTap: onTap,
      isDayOfCurrentMonth: isDayOfCurrentMonth,
      events: events,
      markBuilder: markBuilder,
      borderRadius: borderRadius,
      dayTileMargin: dayTileMargin,
      isSelected: isSelected,
      dayTileBorderColor: dayTileBorderColor,
      showBorder: showBorder,
      isToday: formatDate(dateTime) == formatDate(DateTime.now()),
    );
    return Expanded(
      child: GestureDetector(
          onTap: () {
            if (onTap != null) {
              onTap(dateTime, !isSelected);
            }
          },
          onLongPress: () {},
          child: Container(
            width: double.infinity,
            child: dayTileBuilder != null
                ? dayTileBuilder(dayProps)
                : _DayTile(dayProps),
          )),
    );
  }
}

class DayProps {
  final DateTime dateTime;
  final bool isDayOfCurrentMonth;
  final List<String> events;
  final Function markBuilder;
  final double borderRadius;
  final double dayTileMargin;
  final Function onTap;
  final bool isSelected;
  final Color dayTileBorderColor;
  final bool showBorder;
  final bool isToday;

  DayProps({
    this.dateTime,
    this.isDayOfCurrentMonth,
    this.events,
    this.markBuilder,
    this.borderRadius,
    this.dayTileMargin,
    this.onTap,
    this.isSelected,
    this.dayTileBorderColor,
    this.showBorder,
    this.isToday,
  });
}

class _DayTile extends StatelessWidget {
  final DayProps props;

  _DayTile(this.props);

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: EdgeInsets.all(props.dayTileMargin ?? 3),
      decoration: BoxDecoration(
          border: props.showBorder == true
              ? Border.all(
                  width: 1,
                  color: props.dayTileBorderColor ?? Colors.grey,
                )
              : null,
          borderRadius: BorderRadius.circular(props.borderRadius),
          color: props.isSelected
              ? Theme.of(context).primaryColor
              : props.isToday
                  ? Theme.of(context).primaryColorLight
                  : Colors.transparent),
      child: Stack(
        children: <Widget>[
          Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              Text(
                '${props.dateTime.day}',
                textAlign: TextAlign.center,
                style: TextStyle(
                    color: props.isDayOfCurrentMonth
                        ? Colors.black87
                        : Colors.grey),
              ),
            ],
          ),
          Align(
            alignment: Alignment.bottomCenter,
            child: _EventMarks(props.events, markBuilder: props.markBuilder),
          ),
        ],
      ),
    );
  }
}

class _EventMarks extends StatelessWidget {
  final List<String> events;
  final Function markBuilder;

  _EventMarks(this.events, {this.markBuilder});

  @override
  Widget build(BuildContext context) {
    if (events == null) {
      return Container();
    }

    if (events.length == 0) {
      return Container();
    }

    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: List.generate(events.length, (index) {
        if (this.markBuilder != null) {
          return this.markBuilder(events[index]);
        }

        return _EventMark(Colors.black);
      }),
    );
  }
}

class _EventMark extends StatelessWidget {
  final Color color;

  _EventMark(this.color);

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: EdgeInsets.only(bottom: 5),
      child: Container(
        width: 6,
        height: 6,
        margin: EdgeInsets.symmetric(horizontal: 1),
        color: color,
      ),
    );
  }
}

String formatDate(DateTime date) {
  String day = date.day.toString().padLeft(2, '0');
  String month = date.month.toString().padLeft(2, '0');
  return '${date.year}-$month-$day';
}


example/lib/main.dart
import 'package:example/customized_calendar.dart';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.green,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  Map<String, List<String>> _events;

  Map<String, bool> _selectedDates;

  @override
  void initState() {
    super.initState();
    _events = Map();
    _events[DateFormat('y-M-d').format(DateTime.now())] = ['Hello'];
    _events[DateFormat('y-M-d').format(DateTime(2019, 4, 5))] = ['Hello'];

    _selectedDates = Map<String, bool>();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(            
      appBar: AppBar(
        title: Text('Kalendar Demo'),
      ),
      body: CustomizedCalendar()
    );
  }
}

example/lib/calendar_with_border.dart
import 'package:flutter/material.dart';
import 'package:kalendar/kalendar.dart';

class CalendarWithBorder extends StatefulWidget {
  @override
  _CalendarWithBorderState createState() => _CalendarWithBorderState();
}

class _CalendarWithBorderState extends State<CalendarWithBorder>
    with SingleTickerProviderStateMixin {
  Animation<double> animation;
  AnimationController controller;

  @override
  void initState() {
    super.initState();
    controller =
        AnimationController(duration: Duration(milliseconds: 500), vsync: this);
    animation = Tween<double>(begin: 0, end: 30).animate(controller)
      ..addListener(() {
        setState(() {});
      })
      ..addStatusListener((status) {
        if (status == AnimationStatus.completed) {
          Future.delayed(Duration(seconds: 1), () {
            controller.reverse();
          });
        } else if (status == AnimationStatus.dismissed) {
          Future.delayed(Duration(seconds: 1), () {
            controller.forward();
          });
        }
      });

      controller.forward();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: EdgeInsets.symmetric(horizontal: 4),
      child: Kalendar(
        showBorder: true,
        borderRadius: animation.value,
      ),
    );
  }
}


example/lib/calendar_with_multi_select.dart
import 'dart:collection';

import 'package:flutter/material.dart';
import 'package:kalendar/kalendar.dart';

class CalendarWithMultiSelect extends StatefulWidget {
  @override
  _CalendarWithMultiSelectState createState() =>
      _CalendarWithMultiSelectState();
}

class _CalendarWithMultiSelectState extends State<CalendarWithMultiSelect> {
  final _selectedDates = HashSet<String>();
  Map<String, List<String>> _markedDates;

  @override
  void initState() {    
    super.initState();
    _markedDates = {
      '2019-04-10': [
        'event1'
      ]
    };
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: EdgeInsets.symmetric(horizontal: 8),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          Expanded(
            child: Kalendar(
              showBorder: true,
              borderRadius: 12,
              onTap: (DateTime dateTime, bool isSelected) {
                debugPrint(dateTime.toIso8601String());
                debugPrint('$isSelected');
                setState(() {                  
                  if (isSelected) {
                    _selectedDates.add(formatDate(dateTime));
                  } else {
                    _selectedDates.remove(formatDate(dateTime));
                  }
                });
              },
              selectedDates: _selectedDates,
              markedDates: _markedDates,
            ),
          ),

          Row(
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,            
            children: <Widget>[
              RaisedButton(                                
                onPressed: () {
                  _selectedDates.forEach((date) {
                    if (_markedDates[date] == null) {
                      _markedDates[date] = [];
                    }

                    _markedDates[date].add('event1');
                  });

                  setState(() {});
                },
                child: Text('Mark selected dates'),
              ),
              RaisedButton(                                
                onPressed: () {
                  _selectedDates.forEach((date) {
                    if (_markedDates[date] == null) {
                      _markedDates[date] = [];
                    }

                    _markedDates[date].remove('event1');
                  });

                  setState(() {});
                },
                child: Text('Unmark selected dates'),
              ),
            ],
          ),

          Center(
            child: Padding(
              padding: const EdgeInsets.symmetric(vertical: 16, horizontal: 16),
              child: Text(_selectedDates.toString()),
            ),
          )
          // Text('Hello')
        ],
      ),
    );
  }
}


example/lib/customized_calendar.dart
import 'dart:collection';

import 'package:flutter/material.dart';
import 'package:kalendar/kalendar.dart';

class CustomizedCalendar extends StatefulWidget {
  @override
  _CustomizedCalendarState createState() => _CustomizedCalendarState();
}

class _CustomizedCalendarState extends State<CustomizedCalendar> {
  var _events = Map<String, List<String>>();
  final _selectedDates = HashSet<String>();

  @override
  Widget build(BuildContext context) {  
    return Container(
      padding: EdgeInsets.symmetric(horizontal: 4),
      child: Column(
        children: <Widget>[
          Expanded(
            child: Kalendar(
              selectedDates: _selectedDates,
              markedDates: _events,
              dayTileMargin: 1,
              dayTileBuilder: (DayProps props) {
                return CustomDayTile(props);
              },
              onTap: (DateTime dateTime, bool isSelected) {
                debugPrint(dateTime.toIso8601String());
                debugPrint('$isSelected');
                setState(() {
                  _selectedDates.clear();
                  _selectedDates.add(formatDate(dateTime));                  
                });
              },
            ),
          ),
          
          Wrap(
            alignment: WrapAlignment.center,
            spacing: 16,            
            children: <Widget>[
              RaisedButton(
                onPressed: () {                  
                  _selectedDates.forEach((date) {
                    if (_events[date] == null) {
                      _events[date] = [];
                    }
                    debugPrint('pressed me');
                    _events[date].add('Tennis');
                  });

                  setState(() {});
                },
                child: Text('Tennis'),
              ),
              RaisedButton(                                
                onPressed: () {
                  _selectedDates.forEach((date) {
                    if (_events[date] == null) {
                      _events[date] = [];
                    }

                    _events[date].add('Wedding');
                  });

                  setState(() {});
                },
                child: Text('Wedding'),
              ),
              RaisedButton(                                
                onPressed: () {
                  _selectedDates.forEach((date) {
                    if (_events[date] == null) {
                      _events[date] = [];
                    }

                    _events[date].add('Dentist');
                  });

                  setState(() {});
                },
                child: Text('Dentist'),
              ),
              RaisedButton(                                
                onPressed: () {
                  _selectedDates.forEach((date) {
                    if (_events[date] == null) {
                      _events[date] = [];
                    }

                    _events[date].add('Interview');
                  });

                  setState(() {});
                },
                child: Text('Interview'),
              ),
              RaisedButton(                                
                onPressed: () {
                  _selectedDates.forEach((date) {
                    if (_events[date] == null) {
                      _events[date] = [];
                    }

                    _events[date].add('Blackday');
                  });

                  setState(() {});
                },
                child: Text('Blackday'),
              ),
              RaisedButton(                                
                onPressed: () {
                  _selectedDates.forEach((date) {
                    if (_events[date] == null) {
                      _events[date] = [];
                    }

                    _events[date].add('Holiday');
                  });

                  setState(() {});
                },
                child: Text('Holiday'),
              ),
            ],
          ),

          Center(
            child: Padding(
              padding: const EdgeInsets.symmetric(vertical: 16, horizontal: 16),
              child: Text(_selectedDates.toString()),
            ),
          )
        ],
      ),
    );
  }
}

class CustomDayTile extends StatelessWidget {
  final DayProps props;

  CustomDayTile(this.props);

  @override
  Widget build(BuildContext context) {
    if (props.events != null && props.events[0] == 'Holiday') {
      return Container(
        color: Colors.red,
        child: Center(
          child: Text(
            '${props.dateTime.day}',
            style: TextStyle(
              fontSize: 30,
              fontWeight: FontWeight.bold,
              color: Colors.white,
            ),
          ),
        ),
      );
    }

    if (props.events != null && props.events[0] == 'Blackday') {
      return Container(
        color: Colors.black,
        child: Center(
          child: Text(
            '${props.dateTime.day}',
            style: TextStyle(
              fontSize: 30,
              fontWeight: FontWeight.bold,
              color: Colors.white,
            ),
          ),
        ),
      );
    }

    return Container(
      margin: EdgeInsets.all(props.dayTileMargin ?? 3),
      decoration: BoxDecoration(
        border: Border.all(color: Theme.of(context).primaryColor
            // width: 1,
            // color: props.dayTileBorderColor ?? Colors.grey,
            ),
        // borderRadius: BorderRadius.circular(props.borderRadius),
        color: props.isSelected ? Colors.green : Colors.transparent,
      ),
      child: Stack(
        children: <Widget>[
          Column(
            mainAxisAlignment:
                props.events != null && props.events[0] == 'Tennis'
                    ? MainAxisAlignment.end
                    : MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              Text(
                '${props.dateTime.day}',
                textAlign: TextAlign.center,
                style: TextStyle(
                  fontWeight: FontWeight.bold,
                  color: props.isDayOfCurrentMonth
                      ? Colors.black87
                      : props.isSelected ? Colors.white54 : Colors.grey,
                ),
              ),
            ],
          ),
          _EventMark(props.events),
        ],
      ),
    );
  }
}

class _EventMark extends StatelessWidget {
  final List<String> events;
  _EventMark(this.events);

  @override
  Widget build(BuildContext context) {
    if (events == null) {
      return Container();
    }

    if (events[0] == 'Wedding') {
      return Align(
        alignment: Alignment.bottomCenter,
        child: Container(
          margin: EdgeInsets.all(2),
          padding: EdgeInsets.all(2),
          color: Colors.purple,
          child: Text('Wedding',
              style: Theme.of(context)
                  .textTheme
                  .caption
                  .copyWith(color: Colors.white, fontSize: 11)),
        ),
      );
    }

    if (events[0] == 'Dentist') {
      return Align(
        alignment: Alignment.bottomCenter,
        child: Container(
          margin: EdgeInsets.all(2),
          padding: EdgeInsets.all(2),
          color: Colors.deepOrange,
          child: Text('Dentist',
              style: Theme.of(context)
                  .textTheme
                  .caption
                  .copyWith(color: Colors.white, fontSize: 11)),
        ),
      );
    }

    if (events[0] == 'Tennis') {
      return Container(
        margin: EdgeInsets.all(2),
        padding: EdgeInsets.all(2),
        child: Image.network(
          'https://upload.wikimedia.org/wikipedia/commons/c/c3/P_tennis.png',
          width: 40,
        ),
      );
    }

    if (events[0] == 'Interview') {
      return Align(
        alignment: Alignment.bottomCenter,
        child: Container(
          margin: EdgeInsets.all(2),
          padding: EdgeInsets.all(2),
          color: Colors.black,
          child: Text('Interview',
              style: Theme.of(context)
                  .textTheme
                  .caption
                  .copyWith(color: Colors.white, fontSize: 11)),
        ),
      );
    }

    return Container();
  }
}
